core/tx_pool.go

```diff
@@ -291,38 +308,37 @@ func NewTxPool(config TxPoolConfig, chainconfig *params.ChainConfig, chain block
 func (pool *TxPool) loop() {
 	defer pool.wg.Done()
 
-	// Start the stats reporting and transaction eviction tickers
-	var prevPending, prevQueued, prevStales int
-
-	report := time.NewTicker(statsReportInterval)
+	var (
+		// Start the stats reporting and transaction eviction tickers
+		prevPending, prevQueued, prevStales int
+		report                              = time.NewTicker(statsReportInterval)
+		evict                               = time.NewTicker(evictionInterval)
+		journal                             = time.NewTicker(pool.config.Rejournal)
+		// Track the previous head headers for transaction reorgs
+		head = pool.chain.CurrentBlock()
+	)
 	defer report.Stop()
-
-	evict := time.NewTicker(evictionInterval)
 	defer evict.Stop()
-
-	journal := time.NewTicker(pool.config.Rejournal)
 	defer journal.Stop()
 
-	// Track the previous head headers for transaction reorgs
-	head := pool.chain.CurrentBlock()
-
 	// Keep waiting for and reacting to the various events
 	for {
 		select {
 		// Handle ChainHeadEvent
 		case ev := <-pool.chainHeadCh:
 			if ev.Block != nil {
-				pool.mu.Lock()
 				if pool.chainconfig.IsHomestead(ev.Block.Number()) {
+					pool.mu.Lock()
```

Comment:
This will acquire the lock for every single head event. Is that really needed? 
Also, the semantics change a bit -- since now the `pool.homestead` becomes set synchronously, but the reset happens asyc. Could that have weird effects?

Reply:
Fixed by moving the pool.homestead update into runReorg.

Reply:
Fyi, `homestead` is used when checking TX gas during early validation. I doubt there will be issues with it.